Upgrade Ember CLI to version 2.18 and align with default blueprint

SergeAstapov opened the pull request:
This updates addon source code with RFC 176 Modules API and allows apps to dedupe `node_modules` to re-use babel@6 from the parent app.

However there is one problem with unit tests (two tests fail because of this):
```
not ok 24 Chrome 63.0 - EmberFeatureFlagsInitializer: route has service injected
    ---
        actual: >
            null
        stack: >
            Error: Assertion Failed: Attempting to inject an unknown injection: 'service:Document'
                at new EmberError (http://localhost:7357/assets/vendor.js:24149:25)
                at assert (http://localhost:7357/assets/vendor.js:24392:15)
                at Registry.validateInjections (http://localhost:7357/assets/vendor.js:21436:66)
                at buildInjections (http://localhost:7357/assets/vendor.js:20915:28)
                at injectionsFor (http://localhost:7357/assets/vendor.js:20938:12)
                at FactoryManager.create (http://localhost:7357/assets/vendor.js:21001:30)
                at Proxy.create (http://localhost:7357/assets/vendor.js:20804:20)
                at instantiateFactory (http://localhost:7357/assets/vendor.js:20893:57)
                at _lookup (http://localhost:7357/assets/vendor.js:20851:12)
                at buildInjections (http://localhost:7357/assets/vendor.js:20921:36)
        message: >
            Died on test #1     at http://localhost:7357/assets/tests.js:339:21
                at Array.forEach (<anonymous>)
                at Module.callback (http://localhost:7357/assets/tests.js:338:40)
                at Module.exports (http://localhost:7357/assets/vendor.js:111:32)
                at requireModule (http://localhost:7357/assets/vendor.js:32:18)
                at TestLoader.<anonymous> (http://localhost:7357/assets/test-support.js:10072:11)
                at TestLoader.require (http://localhost:7357/assets/test-support.js:10062:27): Assertion Failed: Attempting to inject an unknown injection: 'service:Document'
```

If service `service:-document` mocked get other error `TypeError: this.document.createElement is not a function`.

@kategengler do you have a suggestion how to fix that problem? Or test an injection in a different way?

kategengler commented:
I'd look into the test setup. In reality, what this addon should do is have breaking change and to no longer inject everywhere by default -- favoring `inject.service`

SergeAstapov commented:
@kategengler I like that suggestion. I can make a PR with that deprecation

kategengler commented:
Great, thanks!

On Wed, Jan 10, 2018 at 11:22 AM, SergeAstapov <notifications@github.com>
wrote:

> @kategengler <https://github.com/kategengler> I like that suggestion. I
> can make a PR with that deprecation
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kategengler/ember-feature-flags/pull/60#issuecomment-356653963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbHOsKCglvlwmzR78zJe2XWlXKgSnTlks5tJOOvgaJpZM4RYbWa>
> .
>

SergeAstapov commented:
@kategengler I've updated `README.md` and removed auto injections functionality. All tests are now green.

Please let me know if you'd like to make any other adjustments to the PR before if can be merged and new version released.

kategengler commented on package.json:
```diff
@@ -8,38 +8,41 @@
-    "ember-cli-qunit": "^1.4.0",
-    "ember-cli-release": "0.2.8",
+    "broccoli-asset-rev": "^2.4.5",
+    "ember-ajax": "^3.0.0",
```
ember-ajax isn't necessary for this addon

kategengler commented on README.md:
```diff
@@ -11,46 +11,60 @@ ember install ember-feature-flags
 ### Usage
 
-This addon injects a property `features` (configurable) into your routes, controllers and components.
+This addon provides a service named `features` (configurable) available for injection into your routes, controller,s components, etc.
```
Typo here by `controller,s`

kategengler commented on README.md:
```diff
@@ -11,46 +11,60 @@ ember install ember-feature-flags
+export default Controller.extend({
+  features: service(),
+
+  plans: computed('features.new-billing-plans', function(){
```
This would need to be `computed('features.newBillingPlans')` ... the feature flags are camelized when turned into bound properties.

kategengler replied:
Perhaps then there should be another example of `isEnabled` use, though...

kategengler commented on README.md:
```diff
@@ -11,46 +11,60 @@ ember install ember-feature-flags
+  features: service(),
+
+  plans: computed('features.new-billing-plans', function(){
+    if (this.get('features').isEnabled('new-billing-plans')){
```
If using the bound property, no reason to use the `isEnabled` api rather than just `get` in the property

kategengler commented on README.md:
````diff
@@ -11,46 +11,60 @@ ember install ember-feature-flags
 {{/if}}
 ```
 
+*NOTE:* `features` service should be injected into respective component or template in order to make above snippet work:
````
can we make this "must be injected into the respective component:"

kategengler commented on README.md:
```diff
@@ -94,7 +108,7 @@ module.exports = function(environment) {
 #### `config.featureFlagsService`
 
-The name of the `features` injection can be customized with the `featureFlagsService` config
+The name of the `features` service can be customized with the `featureFlagsService` config
```
I believe this feature needs to be deprecated when no longer injecting by default.

kategengler commented on README.md:
```diff
@@ -141,9 +155,7 @@ test( "links go to the new homepage", function () {
 If you use `this.features.isEnabled()` in components under integration test, you will need to inject a stub service in your tests. Using ember-qunit 0.4.16 or later, here's how to do this:
```
I believe this whole block is no longer relevant with newer ember testing.

kategengler commented:
Sorry for the merge conflicts, as soon as they are cleared up, I'll merge. 

I'll release 4.0 after this PR and resolving #63
